### PR TITLE
Llama implementation from llama2 branch

### PR DESCRIPTION
**Pull request by marissaposner**

### CODE REVIEW

File: backend/backend/services/graph/graphql_examples.py [hunk 1]
```diff
@@ -14,7 +14,8 @@
                     timestamp
                 }
               }",
-    "Explanation": "Queries the Bored Ape NFT collection and finds the date with the highest daily trade volume in ETH. \n\nThe timestamp is the number of days since the Unix epoch (basically the Unix timestamp, divided by 86400)."
+    "Explanation": "Queries the Bored Ape NFT collection and finds the date with the highest daily trade volume in ETH. \n\nThe timestamp is the number of days since the Unix epoch (basically the Unix timestamp, divided by 86400).",
+    subgraph = "opensea-v2"
```
Comment (gosuto-inzasheru) [inline backend/backend/services/graph/graphql_examples.py]:
```suggestion
    "Subgraph": "opensea-v2"
```

Comment (gosuto-inzasheru) [reply backend/backend/services/graph/graphql_examples.py]: would keep same structure as used for `"Questions"`,`"Explanation"` etc.

File: backend/backend/services/graph/subgraphs.py [hunk 2]
```diff
@@ -12,9 +12,9 @@
 import glob
 import json
 import os
-
 import yaml
 
+PATH = "/subgraphs/subgraphs" #TODO UPDATE
```
Comment (gosuto-inzasheru) [inline backend/backend/services/graph/subgraphs.py]:
```suggestion
```
not in use

File: backend/backend/services/openAI/prompts/__init__.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from .aave_governance import EXAMPLES as aave_governance_examples
+from .opensea import EXAMPLES as opensea_examples
+from .uniswap_v3 import EXAMPLES as uniswapv3_examples
+
+PROTOCOL_TO_PROMPTS = {
+  "aave-governance": aave_governance_examples,
+  "opensea-v2": opensea_examples,
+  "uniswap-v3": uniswapv3_examples,
+
+}
```
Comment (gosuto-inzasheru) [inline backend/backend/services/openAI/prompts/__init__.py:10]: this wont scale, needs a more programmatic approach

File: backend/backend/services/openAI/service.py [hunk 4]
```diff
@@ -11,29 +21,82 @@ def generate_prompt(input):
...
-            max_tokens=2048,
+    def request_gql_for_graph_llama(self, input_query, subgraph):
+        from backend.services.graph.service import GraphService
+        import regex as re
+        graph_service = GraphService(protocol=subgraph)
+        schema = os.path.join(os.getcwdb().decode("utf-8"), graph_service.subgraph.schema_file_location)
+        mappings = os.path.join(os.getcwdb().decode("utf-8"), "/subgraphs/subgraphs/{}/src/".format(graph_service.subgraph.deployments['base']))
+        examples = os.getcwdb().decode("utf-8")+ "/backend/services/graph/graphql_examples.py"
+        #set recursive = True for case of uniswap etc where there are more sub directories
+        documents = SimpleDirectoryReader(input_files=[examples], recursive=True).load_data()
```
Comment (gosuto-inzasheru) [inline backend/backend/services/openAI/service.py]: `schema` and `mappings` are not taken into account at all atm

File: backend/backend/services/openAI/service.py [hunk 5]
```diff
@@ -11,29 +21,82 @@ def generate_prompt(input):
...
+        # print("documents", documents)
+        # save to disk
+        # index.save_to_disk('index.json')
+        # load from disk
+        # index = GPTSimpleVectorIndex.load_from_disk('index.json')
+
+        # define LLM
+        llm_predictor = LLMPredictor(llm=OpenAI(temperature=0, model_name="text-davinci-003"))
```
Comment (gosuto-inzasheru) [inline backend/backend/services/openAI/service.py:49]: is temperature `0` correct?

File: backend/backend/services/openAI/service.py [hunk 6]
```diff
@@ -11,29 +21,82 @@ def generate_prompt(input):
...
-    def request_gql_for_graph(self, input_query, subgraph, generic=False):
-        prompt = input_query if generic else generate_prompt(input_query)
-        response = openai.Completion.create(
-            model="text-davinci-003",
-            prompt=generate_prompt(input_query),
-            temperature=0.6,
-            max_tokens=2048,
+    def request_gql_for_graph_llama(self, input_query, subgraph):
+        from backend.services.graph.service import GraphService
+        import regex as re
```
Comment (gosuto-inzasheru) [inline backend/backend/services/openAI/service.py]: not being used

File: backend/backend/services/openAI/service.py [hunk 7]
```diff
@@ -11,29 +21,82 @@ def generate_prompt(input):
...
+    #         max_tokens=2048,
+    #     )
+    #     openai_result = response.choices[0].text
+    #     print("==========openai response:==========\n", openai_result)
+    #     # strip any unnecessary text prepended and/or postpended to the gql query
+    #     openai_result = openai_result[
+    #         openai_result.find("{") : openai_result.rfind("}") + 1
+    #     ]
+    #     print("==========openai response (formatted):==========\n", openai_result)
+    #     return openai_result
```
Comment (gosuto-inzasheru) [inline backend/backend/services/openAI/service.py]: would suggest to keep this in the implementation, sometimes gpt still gives weird unparsable graphql code back. stripping like this prevents the code from not being readable by the graph's api

**Review by gosuto-inzasheru — COMMENTED** [review]
do we need `llamaindexttest.py`?

**Comment by gosuto-inzasheru** [comment]
commented @townie 

also has a merge conflict

File: backend/backend/services/openAI/service.py (hunk 4 above)
Comment (gosuto-inzasheru) [reply backend/backend/services/openAI/service.py]: solved in #53

File: backend/backend/services/openAI/service.py (hunk 6 above)
Comment (gosuto-inzasheru) [reply backend/backend/services/openAI/service.py]: solved in #53

File: backend/backend/services/openAI/prompts/__init__.py (hunk 3 above)
Comment (townie) [reply backend/backend/services/openAI/prompts/__init__.py:10]: Agreed. Scale is not a concern for hackathon.

File: backend/backend/services/openAI/service.py (hunk 5 above)
Comment (townie) [reply backend/backend/services/openAI/service.py:49]: Yes. We dont want to much creativity from. 0 == less, 1 == more creativity

File: backend/backend/services/openAI/prompts/__init__.py (hunk 3 above)
Comment (townie) [reply backend/backend/services/openAI/prompts/__init__.py:10]: I think it was a dynamic importing system at one point

**Review by townie — APPROVED** [review]

File: backend/backend/services/openAI/service.py (hunk 7 above)
Comment (townie) [reply backend/backend/services/openAI/service.py]: I was pondering about how to parse the gql, we could have tossed the schema into the client and validated a request on request.

Comment (townie) [reply backend/backend/services/openAI/service.py]: Maybe a formatter in the future

File: backend/backend/services/graph/graphql_examples.py (hunk 1 above)
Comment (townie) [reply backend/backend/services/graph/graphql_examples.py]: This file should be deleted.

**Pull request merged**